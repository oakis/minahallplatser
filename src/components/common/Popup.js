import React, { PureComponent } from 'react';
import { View, Animated, Easing, TouchableOpacity, ScrollView } from 'react-native';
import { Button } from './';
import { component } from '../style/component';

export class Popup extends PureComponent {

    constructor(props) {
        super(props);
        this.state = {
            opacity: new Animated.Value(0),
            isVisible: this.props.isVisible,
            hidden: true,
            transitioning: false,
            scale: new Animated.Value(0.5)
        };
    }

    componentWillReceiveProps(nextProps) {
        if (nextProps.isVisible && !this.state.isVisible) {
            this.setState({ isVisible: true, hidden: false, transitioning: true });
            Animated.timing(this.state.scale, {
                toValue: 1,
                easing: Easing.elastic(),
                duration: 500,
                useNativeDriver: true
            }).start();
            Animated.timing(this.state.opacity, {
                toValue: 1,
                easing: Easing.ease,
                duration: 250,
                useNativeDriver: true
            }).start(() => {
                this.setState({ transitioning: false });
            });
        } else if (!nextProps.isVisible && this.state.isVisible) {
            this.setState({ isVisible: false, transitioning: true });
            Animated.timing(this.state.scale, {
                toValue: 0,
                easing: Easing.inOut(Easing.back()),
                duration: 500,
                useNativeDriver: true
            }).start();
            Animated.timing(this.state.opacity, {
                toValue: 0,
                easing: Easing.linear,
                duration: 250,
                useNativeDriver: true
            }).start(() => {
                this.setState({ hidden: true, transitioning: false });
            });
        }
    }

    render() {
        const { children, onPress } = this.props;
        const { opacity, hidden, transitioning, scale } = this.state;
        return (
            <View style={[component.popup.container, { height: hidden ? 0 : '100%', width: hidden ? 0 : '100%', backgroundColor: 'transparent' }]}>
                <TouchableOpacity activeOpacity={1} disabled={transitioning} onPress={onPress} style={{ position: 'absolute', zIndex: 1, height: hidden ? 0 : '100%', width: hidden ? 0 : '100%' }}>
                    <Animated.View style={[component.popup.container, { opacity }]} />
                </TouchableOpacity>
                <Animated.View style={[component.popup.content, { opacity, transform: [{ scale }] }]}>
                    <ScrollView>
                        {children}
                        <Button label="StÃ¤ng" uppercase color={'primary'} onPress={onPress} />
                    </ScrollView>
                </Animated.View>
            </View>
        );
    }
}
