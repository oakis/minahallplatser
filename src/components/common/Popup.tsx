import React, {useEffect, useRef, useState} from 'react';
import {View, TouchableOpacity, StyleProp, ViewStyle} from 'react-native';
import Animated, {EasingNode} from 'react-native-reanimated';
import {Button} from '@common';
import {metrics, component, colors} from '@style';

const duration = 160;

interface PopupProps {
  isVisible: boolean;
  onPress: () => void;
  children: React.ReactNode;
}

export const Popup = (props: PopupProps): JSX.Element => {
  const [hidden, setHidden] = useState(true);

  const animateValue = useRef(new Animated.Value(0)).current;

  useEffect(() => {
    if (props.isVisible) {
      setHidden(false);
      Animated.timing(animateValue, {
        toValue: 1,
        easing: EasingNode.elastic(),
        duration,
      }).start();
      Animated.timing(animateValue, {
        toValue: 1,
        easing: EasingNode.ease,
        duration,
      }).start();
    } else if (!props.isVisible) {
      Animated.timing(animateValue, {
        toValue: 0,
        easing: EasingNode.ease,
        duration,
      }).start();
      Animated.timing(animateValue, {
        toValue: 0,
        easing: EasingNode.ease,
        duration,
      }).start();
      setTimeout(() => {
        setHidden(true);
      }, duration);
    }
  }, [animateValue, props.isVisible]);

  return (
    <View
      style={
        [
          component.popup.container,
          {
            height: hidden ? 0 : '100%',
            width: hidden ? 0 : '100%',
            backgroundColor: 'transparent',
          },
        ] as StyleProp<ViewStyle>
      }>
      <TouchableOpacity
        activeOpacity={1}
        onPress={props.onPress}
        style={{
          position: 'absolute',
          zIndex: 1,
          height: hidden ? 0 : '100%',
          width: hidden ? 0 : '100%',
        }}>
        <Animated.View
          style={[component.popup.container, {opacity: animateValue}]}
        />
      </TouchableOpacity>
      <Animated.ScrollView
        style={[
          component.popup.content,
          {opacity: animateValue, transform: [{scale: animateValue}]},
        ]}
        scrollEnabled
        keyboardShouldPersistTaps="always">
        {props.children}
        <Button
          label="StÃ¤ng"
          uppercase
          color={colors.primary}
          onPress={props.onPress}
          style={{marginBottom: metrics.margin.xl}}
        />
      </Animated.ScrollView>
    </View>
  );
};

export default Popup;
