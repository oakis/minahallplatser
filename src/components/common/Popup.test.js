import React from 'react';
import { Animated } from 'react-native';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Popup } from './Popup';

Animated.Value = () => (0);

it('should match snapshot', () => {
    const wrapper = shallow(<Popup />);
    expect(toJson(wrapper)).toMatchSnapshot();
});

it('should match snapshot', () => {
    const wrapper = shallow(<Popup />);
    wrapper.setProps({ isVisible: true });
    expect(toJson(wrapper)).toMatchSnapshot();
});

it('should match snapshot', () => {
    const wrapper = shallow(<Popup />);
    wrapper.setState({ isVisible: true });
    wrapper.setProps({ isVisible: false });
    expect(toJson(wrapper)).toMatchSnapshot();
});
