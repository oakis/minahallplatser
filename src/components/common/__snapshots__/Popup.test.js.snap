// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(0, 0, 0, 0.4)",
        "height": "100%",
        "justifyContent": "center",
        "position": "absolute",
        "width": "100%",
        "zIndex": 1,
      },
      Object {
        "backgroundColor": "transparent",
        "height": 0,
        "width": 0,
      },
    ]
  }
>
  <TouchableOpacity
    activeOpacity={1}
    style={
      Object {
        "height": 0,
        "position": "absolute",
        "width": 0,
        "zIndex": 1,
      }
    }
  >
    <Animated.View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "rgba(0, 0, 0, 0.4)",
            "height": "100%",
            "justifyContent": "center",
            "position": "absolute",
            "width": "100%",
            "zIndex": 1,
          },
          Object {
            "opacity": mockConstructor {},
          },
        ]
      }
    />
  </TouchableOpacity>
  <Component
    keyboardShouldPersistTaps="always"
    scrollEnabled={true}
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "borderRadius": 5,
          "borderWidth": 2,
          "elevation": 10,
          "maxHeight": 1329.2307692307693,
          "minHeight": 250,
          "padding": 10,
          "position": "absolute",
          "width": "85%",
          "zIndex": 2,
        },
        Object {
          "opacity": mockConstructor {},
          "transform": Array [
            Object {
              "scale": mockConstructor {},
            },
          ],
        },
      ]
    }
  >
    <Button
      color="primary"
      label="Stäng"
      style={
        Object {
          "marginBottom": 30,
        }
      }
      uppercase={true}
    />
  </Component>
</View>
`;

exports[`should match snapshot 2`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(0, 0, 0, 0.4)",
        "height": "100%",
        "justifyContent": "center",
        "position": "absolute",
        "width": "100%",
        "zIndex": 1,
      },
      Object {
        "backgroundColor": "transparent",
        "height": "100%",
        "width": "100%",
      },
    ]
  }
>
  <TouchableOpacity
    activeOpacity={1}
    style={
      Object {
        "height": "100%",
        "position": "absolute",
        "width": "100%",
        "zIndex": 1,
      }
    }
  >
    <Animated.View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "rgba(0, 0, 0, 0.4)",
            "height": "100%",
            "justifyContent": "center",
            "position": "absolute",
            "width": "100%",
            "zIndex": 1,
          },
          Object {
            "opacity": mockConstructor {},
          },
        ]
      }
    />
  </TouchableOpacity>
  <Component
    keyboardShouldPersistTaps="always"
    scrollEnabled={true}
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "borderRadius": 5,
          "borderWidth": 2,
          "elevation": 10,
          "maxHeight": 1329.2307692307693,
          "minHeight": 250,
          "padding": 10,
          "position": "absolute",
          "width": "85%",
          "zIndex": 2,
        },
        Object {
          "opacity": mockConstructor {},
          "transform": Array [
            Object {
              "scale": mockConstructor {},
            },
          ],
        },
      ]
    }
  >
    <Button
      color="primary"
      label="Stäng"
      style={
        Object {
          "marginBottom": 30,
        }
      }
      uppercase={true}
    />
  </Component>
</View>
`;

exports[`should match snapshot 3`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(0, 0, 0, 0.4)",
        "height": "100%",
        "justifyContent": "center",
        "position": "absolute",
        "width": "100%",
        "zIndex": 1,
      },
      Object {
        "backgroundColor": "transparent",
        "height": 0,
        "width": 0,
      },
    ]
  }
>
  <TouchableOpacity
    activeOpacity={1}
    style={
      Object {
        "height": 0,
        "position": "absolute",
        "width": 0,
        "zIndex": 1,
      }
    }
  >
    <Animated.View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "rgba(0, 0, 0, 0.4)",
            "height": "100%",
            "justifyContent": "center",
            "position": "absolute",
            "width": "100%",
            "zIndex": 1,
          },
          Object {
            "opacity": mockConstructor {},
          },
        ]
      }
    />
  </TouchableOpacity>
  <Component
    keyboardShouldPersistTaps="always"
    scrollEnabled={true}
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "borderRadius": 5,
          "borderWidth": 2,
          "elevation": 10,
          "maxHeight": 1329.2307692307693,
          "minHeight": 250,
          "padding": 10,
          "position": "absolute",
          "width": "85%",
          "zIndex": 2,
        },
        Object {
          "opacity": mockConstructor {},
          "transform": Array [
            Object {
              "scale": mockConstructor {},
            },
          ],
        },
      ]
    }
  >
    <Button
      color="primary"
      label="Stäng"
      style={
        Object {
          "marginBottom": 30,
        }
      }
      uppercase={true}
    />
  </Component>
</View>
`;
